File: ios/Classes/AiEdgePlugin.swift
import Flutter
import MediaPipeTasksGenAI

private enum CallError: Error {
    case unknown(String)
    case createModel(String)
    case createSession(String)
    case addQueryChunk(String)
    case addImage(String)
    case generateResponse(String)
    case generateResponseAsync(String)

    var code: String {
        switch self {
        case .unknown: "UNKNOWN_ERROR"
        case .createModel: "CREATE_MODEL_ERROR"
        case .createSession: "CREATE_SESSION_ERROR"
        case .addQueryChunk: "ADD_QUERY_CHUNK_ERROR"
        case .addImage: "ADD_IMAGE_ERROR"
        case .generateResponse: "GENERATE_RESPONSE_ERROR"
        case .generateResponseAsync: "GENERATE_RESPONSE_ASYNC_ERROR"
        }
    }

    var message: String {
        switch self {
        case .unknown(let msg),
            .createModel(let msg),
            .createSession(let msg),
            .addQueryChunk(let msg),
            .addImage(let msg),
            .generateResponse(let msg),
            .generateResponseAsync(let msg):
            msg
        }
    }

    func toFlutterError() -> FlutterError {
        FlutterError(code: code, message: message, details: nil)
    }
}

public class AiEdgePlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let methodChannel = FlutterMethodChannel(
            name: "ai_edge/methods",
            binaryMessenger: registrar.messenger()
        )
        let eventChannel = FlutterEventChannel(
            name: "ai_edge/events",
            binaryMessenger: registrar.messenger()
        )

        let instance = AiEdgePlugin()
        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        eventChannel.setStreamHandler(instance)
    }

    private var eventSink: FlutterEventSink?
    private var inferenceModel: InferenceModel?
    private var session: InferenceSession?

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        Task { @MainActor in
            do {
                let response =
                    switch call.method {
                    case "createModel":
                        try await handleCreateModel(call)
                    case "createSession":
                        try await handleCreateSession(call)
                    case "addQueryChunk":
                        try await handleAddQueryChunk(call)
                    case "addImage":
                        try await handleAddImage(call)
                    case "generateResponse":
                        try await handleGenerateResponse(call)
                    case "generateResponseAsync":
                        try await handleGenerateResponseAsync(call)
                    case "close":
                        await handleClose()
                    default:
                        FlutterMethodNotImplemented
                    }
                result(response as? String)
            } catch let error as CallError {
                result(error.toFlutterError())
            } catch {
                let error = CallError.unknown(error.localizedDescription)
                result(error.toFlutterError())
            }
        }
    }

    private func handleCreateModel(_ call: FlutterMethodCall) async throws {
        guard let args = call.arguments as? [String: Any],
            let modelPath = args["modelPath"] as? String,
            let maxTokens = args["maxTokens"] as? Int
        else {
            throw CallError.createModel("Missing modelPath or maxTokens")
        }

        let loraRanks = args["loraRanks"] as? [Int]
        let maxNumImages = args["maxNumImages"] as? Int

        do {
            inferenceModel = try InferenceModel(
                modelPath: modelPath,
                maxTokens: maxTokens,
                supportedLoraRanks: loraRanks,
                maxNumImages: maxNumImages
            )
        } catch {
            throw CallError.createModel(error.localizedDescription)
        }
    }

    private func handleCreateSession(_ call: FlutterMethodCall) async throws {
        guard let inference = inferenceModel?.inference else {
            throw CallError.createSession("Inference model is not created")
        }

        guard let args = call.arguments as? [String: Any],
            let temperature = args["temperature"] as? Double,
            let randomSeed = args["randomSeed"] as? Int,
            let topK = args["topK"] as? Int
        else {
            throw CallError.createSession("Missing temperature, randomSeed or topK")
        }

        let topP = args["topP"] as? Double
        let loraPath = args["loraPath"] as? String
        let enableVisionModality = args["enableVisionModality"] as? Bool

        do {
            session = try InferenceSession(
                inference: inference,
                temperature: Float(temperature),
                randomSeed: randomSeed,
                topk: topK,
                topP: topP,
                loraPath: loraPath,
                enableVisionModality: enableVisionModality
            )
        } catch {
            throw CallError.createSession(error.localizedDescription)
        }
    }

    private func handleAddQueryChunk(_ call: FlutterMethodCall) async throws {
        guard let session else {
            throw CallError.addQueryChunk("Session not created")
        }

        guard let args = call.arguments as? [String: Any],
            let prompt = args["prompt"] as? String
        else {
            throw CallError.addQueryChunk("Missing prompt")
        }

        do {
            try session.addQueryChunk(prompt: prompt)
        } catch {
            throw CallError.addQueryChunk(error.localizedDescription)
        }
    }

    private func handleAddImage(_ call: FlutterMethodCall) async throws {
        guard let session else {
            throw CallError.addImage("Session not created")
        }

        guard let args = call.arguments as? [String: Any],
            let imageBytes = args["imageBytes"] as? FlutterStandardTypedData
        else {
            throw CallError.addImage("Missing imageBytes")
        }

        guard let cgImage = UIImage(data: imageBytes.data)?.cgImage else {
            throw CallError.addImage("Could not create image from data")
        }

        do {
            try session.addImage(image: cgImage)
        } catch {
            throw CallError.addImage(error.localizedDescription)
        }
    }

    private func handleGenerateResponse(_ call: FlutterMethodCall) async throws -> String {
        guard let session else {
            throw CallError.generateResponse("Session not created")
        }
        let args = call.arguments as? [String: String]

        do {
            return try session.generateResponse(prompt: args?["prompt"])
        } catch {
            throw CallError.generateResponse(error.localizedDescription)
        }
    }

    private func handleGenerateResponseAsync(_ call: FlutterMethodCall) async throws {
        guard let session else {
            throw CallError.generateResponseAsync("Session not created")
        }

        guard let eventSink else {
            throw CallError.generateResponseAsync("Event sink not created")
        }

        @MainActor
        func sink(_ event: Any) {
            eventSink(event)
        }

        do {
            let args = call.arguments as? [String: String]
            let stream = try session.generateResponseAsync(prompt: args?["prompt"])
            Task.detached {
                do {
                    for try await token in stream {
                        await sink(["partialResult": token, "done": false])
                    }
                    await sink(["partialResult": "", "done": true])
                    await sink(FlutterEndOfEventStream)
                } catch {
                    let error = CallError.generateResponseAsync(error.localizedDescription)
                    await sink(error.toFlutterError())
                }
            }
        } catch {
            throw CallError.generateResponseAsync(error.localizedDescription)
        }
    }

    private func handleClose() async -> Any? {
        session = nil
        inferenceModel = nil
        return nil
    }
}

// MARK: - FlutterStreamHandler
extension AiEdgePlugin: FlutterStreamHandler {
    public func onListen(
        withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink
    ) -> FlutterError? {
        eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }
}

File: android/src/main/kotlin/jp/kyoheig3/ai_edge/AiEdgePlugin.kt
package jp.kyoheig3.ai_edge

import android.content.Context
import android.util.Log
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.*

sealed class CallError(override val message: String) : Exception(message) {
    data class Unknown(val msg: String) : CallError(msg)
    data class CreateModel(val msg: String) : CallError(msg)
    data class CreateSession(val msg: String) : CallError(msg)
    data class AddQueryChunk(val msg: String) : CallError(msg)
    data class AddImage(val msg: String) : CallError(msg)
    data class GenerateResponse(val msg: String) : CallError(msg)
    data class GenerateResponseAsync(val msg: String) : CallError(msg)
    
    val code: String
        get() = when (this) {
            is Unknown -> "UNKNOWN_ERROR"
            is CreateModel -> "CREATE_MODEL_ERROR"
            is CreateSession -> "CREATE_SESSION_ERROR"
            is AddQueryChunk -> "ADD_QUERY_CHUNK_ERROR"
            is AddImage -> "ADD_IMAGE_ERROR"
            is GenerateResponse -> "GENERATE_RESPONSE_ERROR"
            is GenerateResponseAsync -> "GENERATE_RESPONSE_ASYNC_ERROR"
        }
}

class AiEdgePlugin: FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {
    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private lateinit var context: Context
    private val scope = CoroutineScope(Dispatchers.IO)
    private var eventSink: EventChannel.EventSink? = null
    private var inferenceModel: InferenceModel? = null
    private var session: InferenceSession? = null

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        methodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "ai_edge/methods")
        methodChannel.setMethodCallHandler(this)
        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "ai_edge/events")
        eventChannel.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        scope.launch {
            try {
                val response = when (call.method) {
                    "createModel" -> handleCreateModel(call)
                    "createSession" -> handleCreateSession(call)
                    "addQueryChunk" -> handleAddQueryChunk(call)
                    "addImage" -> handleAddImage(call)
                    "generateResponse" -> handleGenerateResponse(call)
                    "generateResponseAsync" -> handleGenerateResponseAsync(call)
                    "close" -> handleClose()
                    else -> throw NotImplementedError("Method not implemented: ${call.method}")
                }
                launch(Dispatchers.Main) {
                    result.success(response as? String)
                }
            } catch (_: NotImplementedError) {
                launch(Dispatchers.Main) {
                    result.notImplemented()
                }
            } catch (e: CallError) {
                launch(Dispatchers.Main) {
                    result.error(e.code, e.message, e.stackTraceToString())
                }
            } catch (e: Exception) {
                launch(Dispatchers.Main) {
                    val error = CallError.Unknown(e.message ?: "Unknown error")
                    result.error(error.code, error.message, e.stackTraceToString())
                }
            }
        }
    }

    private fun handleCreateModel(call: MethodCall) {
        val modelPath = call.argument<String>("modelPath")
            ?: throw CallError.CreateModel("Missing modelPath")
        val maxTokens = (call.argument<Number>("maxTokens")
            ?: throw CallError.CreateModel("Missing maxTokens")).toInt()

        val loraRanks = call.argument<List<Number>>("loraRanks")?.map { it.toInt() }
        val maxNumImages = call.argument<Number>("maxNumImages")?.toInt()

        val preferredBackend = call.argument<Number>("preferredBackend")?.toInt()?.let {
            PreferredBackend.entries.getOrNull(it)
        }

        try {
            inferenceModel?.close()
            inferenceModel = InferenceModel(
                context,
                modelPath,
                maxTokens,
                loraRanks,
                maxNumImages,
                preferredBackend,
            )
        } catch (e: Exception) {
            throw CallError.CreateModel(e.message ?: e.toString())
        }
    }

    private fun handleCreateSession(call: MethodCall) {
        val inference = inferenceModel?.inference
            ?: throw CallError.CreateSession("Inference model is not created")

        val temperature = call.argument<Number>("temperature")?.toFloat()
            ?: throw CallError.CreateSession("Missing temperature")
        val randomSeed = call.argument<Number>("randomSeed")?.toInt()
            ?: throw CallError.CreateSession("Missing randomSeed")
        val topK = call.argument<Number>("topK")?.toInt()
            ?: throw CallError.CreateSession("Missing topK")

        val topP = call.argument<Number>("topP")?.toFloat()
        val loraPath = call.argument<String>("loraPath")
        val enableVisionModality = call.argument<Boolean>("enableVisionModality")

        try {
            session?.close()
            session = InferenceSession(
                inference,
                temperature,
                randomSeed,
                topK,
                topP,
                loraPath,
                enableVisionModality
            )
        } catch (e: Exception) {
            throw CallError.CreateSession(e.message ?: e.toString())
        }
    }

    private fun handleAddQueryChunk(call: MethodCall) {
        val session = session
            ?: throw CallError.AddQueryChunk("Session not created")
        val prompt = call.argument<String>("prompt") 
            ?: throw CallError.AddQueryChunk("Missing prompt")

        try {
            session.addQueryChunk(prompt)
        } catch (e: Exception) {
            throw CallError.AddQueryChunk(e.message ?: e.toString())
        }
    }

    private fun handleAddImage(call: MethodCall) {
        val session = session 
            ?: throw CallError.AddImage("Session not created")
        val imageBytes = call.argument<ByteArray>("imageBytes") 
            ?: throw CallError.AddImage("Missing imageBytes")

        try {
            session.addImage(imageBytes)
        } catch (e: Exception) {
            throw CallError.AddImage(e.message ?: e.toString())
        }
    }

    private fun handleGenerateResponse(call: MethodCall): String {
        val session = session 
            ?: throw CallError.GenerateResponse("Session not created")

        return try {
            session.generateResponse(call.argument("prompt"))
        } catch (e: Exception) {
            throw CallError.GenerateResponse(e.message ?: e.toString())
        }
    }

    private fun handleGenerateResponseAsync(call: MethodCall) {
        val session = session 
            ?: throw CallError.GenerateResponseAsync("Session not created")
        val eventSink = eventSink
            ?: throw CallError.GenerateResponseAsync("Event sink not created")

        session.generateResponseAsync(call.argument("prompt")) { result, done ->
            val payload = mapOf("partialResult" to result, "done" to done)
            scope.launch(Dispatchers.Main) {
                eventSink.success(payload)
                if (done) {
                    eventSink.endOfStream()
                }
            }
        }
    }

    private fun handleClose() {
        try {
            session?.close()
        } catch (e: IllegalStateException) {
            // Session is still processing, ignore the error
            Log.w("AiEdgePlugin", "Session close failed: ${e.message}")
        }
        session = null

        inferenceModel?.close()
        inferenceModel = null
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }
}